### PR TITLE
Prometheus middleware

### LINKED ISSUE #4: Prometheus Middleware
Using the middleware function that's available now (which may including upgrading/enhancing that functionality), create something like a `PrometheusMiddleware` that we can add to a host, e.g.

```
host::add_middleware(PrometheusMiddleware::new(/* ... config params .. */));
```
At a minimum, I would think that we would want to keep counters on the following information:

* capability call count (total and per capability)  - count of invocations requested by the actor of the host runtime
* actor call count (per actor and total) - count of invocations made by capabilities/host on each actor
* capability call time (per capability and total) - maintain a running avg of how long it takes to invoke a capability 
* actor call time (per actor and total) - maintain a running avg of how long it takes to call an actor

This middleware should also support the following:
* Push model - emit the prometheus data to a push gateway
* Pull model - start an HTTP server that allows for pull scraping of data
* Actor SDK should allow developers to maintain their own counters for feature-level metrics, e.g. `ctx.metrics().xxx()`.
   * Existence or non-existence of metrics middleware should not require a developer to recompile an actor, so all these functions should appear to have no effect when middleware is missing.

As mentioned, it is quite likely that we'll need to modify how some of the middleware system works in order to allow this middleware to get all the information it needs.

### PR DESCRIPTION
**Pull request by andrisak**
This is a WIP/draft implementation of push and pull support for Prometheus. Would like to validate the approach before spending too much time on it.

Fixes #4.

### CODE REVIEW

**Review by autodidaddict — CHANGES_REQUESTED** [review]
This looks great so far, thanks so much for the contribution! I think there's a couple of organizational things that will help make this change more friendly to other library consumers:

1) Let's move the `src/prometheus_middleware` module to `src/middleware/prometheus.rs`. Then `middleware.rs` becomes `src/middleware/mod.rs` and the `PrometheusMiddleware` struct can be found at `middleware::prometheus::PrometheusMiddleware` This helps make it easier to do the work for item 2
2) Create a feature flag `prometheus` that only includes the prometheus module and the required dependencies if this feature is enabled. This will make it so that taking on the prometheus dependencies is an opt-in choice made by library consumers.
3) Finish up the TODO items

Is there a docker image we can run to locally test out the push gateway version? If so, can we get that in a README or code comment?

**Comment by andrisak** [comment]
Good points, I'll make those adjustments.

There is one docker image for the [pushgateway](https://hub.docker.com/r/prom/pushgateway) and one for [Prometheus](https://hub.docker.com/r/prom/prometheus). I have a config file for Prometheus and use docker-compose to run the images. I will try to find a place to put those files.

File: src/prometheus_middleware.rs [hunk 1]
```diff
@@ -61,33 +85,35 @@ impl PrometheusMiddleware {
...
 
-            // need a separate thread for pushing metrics because the reqwest sync client
-            // is used in the prometheus library. reqwest sync creates a tokio runtime internally
-            // so the tokio runtime created for the metrics server can't be reused.
-            let thread_handle = std::thread::spawn(move || {
-                push_metrics(addr, metrics_push_kill_switch_rx);
-            });
+                // need a separate thread for pushing metrics because the reqwest sync client
+                // is used in the prometheus library. reqwest sync creates a tokio runtime internally
+                // so the tokio runtime created for the metrics server can't be reused.
```
Comment (autodidaddict) [inline src/prometheus_middleware.rs]: This statement sums up so much of what I don't like about the current state of async runtimes in Rust. Those kinds of internal details (like reqwest creating a tokio runtime as a block point) should never force consumers to make these kinds of decisions/refactors.

**Review by autodidaddict — CHANGES_REQUESTED** [review]
Really enjoying watching this come together. Once the refactoring has been done to move the prometheus stuff to a separate module that is then hidden behind a feature flag, let us know when this PR is no longer a WIP.

I'd also love a "setup guide" somewhere in a README that shows me which docker images I can run and which wascc sample to run to watch the stats being gathered in prometheus.

Thanks

**Comment by autodidaddict** [comment]
Also looks like the tests are failing, you can see the log in the details of the failed PR build.

**Comment by andrisak** [comment]
I've done the refactoring and many of the todos. However, I haven't implemented the timing
metrics. The `InvocationResponse` the middleware gets called with doesn't seem to include anything that can connect it to the corresponding `Invocation`. One way to solve this would be adding a uuid (maybe using [this crate](https://docs.rs/uuid/0.8.1/uuid/)?) to both these structs so that requests and replies can be tracked. Do you have any thoughts about this?

Also added an example showing how the middleware can be used. Unfortunately, it doesn't run on my machine (Mac). Built the http_server_provider project locally and changed the path to the built dylib-file (as described [here](https://wascc.dev/tutorials/first-actor/create_host/)). But when running the example, there is an error:

`Error: Error(Plugin(DlOpen { desc: "dlopen(../../http-server-provider/target/debug/libwascc_httpsrv.dylib, 2): image not found" }))`

Any ideas?

**Comment by autodidaddict** [comment]
If you are having trouble with dynamically loading the capability providers
you can instantiate them as direct dependencies, just use the static_plugin
feature in the cargo dependencies

that said, the `image not found` error is usually a failure to find the actual file. Make sure it's been copied to the right location.

**Comment by autodidaddict** [comment]
Added unique IDs in this PR: https://github.com/wascc/wascc-host/pull/32

**Comment by autodidaddict** [comment]
@andrisak let me know if the recent changes from master were enough to close that gap on the timing generation.

**Comment by andrisak** [comment]
@autodidaddict had a look at it yesterday and it looks like it did the trick. I still have some implementation and testing left though. I hope to be able to push something during the coming couple of days.

**Comment by andrisak** [comment]
Now a first version of invocation average times are implemented too. There are still some things too think about:

1. Right now it counts the number invocations of an actor or capability, not the number of invocations of a particular operation. Average invocation time is also only per capability or actor, not per operation. Is this what we want?
2. It measures the invocation time in milliseconds. Is microseconds also relevant? Should it be configurable?
3. It uses a sliding window and calculates the average for the 100 last invocations. Maybe the window size should be made configurable too?

**Comment by autodidaddict** [comment]
1. Ideally I'd love to be able to choose which one I want to look at. Can we have counters for actor ID, capability ID, and operation name? 
2. I think milliseconds resolution is just fine for now
3. Can we make it configurable but default it to 100

I don't know if it's there and I just can't find it, but is there a set of instructions on how to test this middleware with docker instances of prometheus (and optionally grafana, etc)? I'd love to be able to play with this branch but my prometheus-fu is old and rusty.

**Comment by andrisak** [comment]
Sounds good, I'll try to implement those counters and averages.

There are some comments in the top of `src/middleware/prometheus.rs` that shows an example Prometheus config and how to run the containers with `docker-compose`. Could that be what you are looking for?

**Comment by autodidaddict** [comment]
Yep, the rustdocs in that module are perfect. I'll try and use those to play around and generate some data.

**Comment by andrisak** [comment]
Good. I just pushed two commits from tonight that should generate counts and averages on operation level as well. It needs some cleaning up and the tests also needs to be adjusted, but I think it might give you better data.

**Comment by autodidaddict** [comment]
looks like the example doesn't compile:
```
error[E0063]: missing field `moving_average_window_size` in initializer of `wascc_host::middleware::prometheus::PrometheusConfig`
  --> examples/prometheus_middleware.rs:30:22
   |
30 |         let config = PrometheusConfig {
   |                      ^^^^^^^^^^^^^^^^ missing `moving_average_window_size`

error: aborting due to previous error
```

**Comment by autodidaddict** [comment]
When I run the example, I get these two errors:
```
[2020-05-28T12:44:30Z ERROR wascc_host::middleware::prometheus] Error registering gauge 'wascc:http_server_default_average_inv_time': Error: descriptor Desc { fq_name: "wascc:http_server_default_average_inv_time", help: "Average time to invoke operation \'BindActor\' on capability \'wascc:http_server\' with binding \'default\'", const_label_pairs: [], variable_labels: [], id: 2460217315795522843, dim_hash: 4187395188162649320 } already exists with the same fully-qualified name and const label values
```
and then after the second actor binds:
```
[2020-05-28T12:44:30Z ERROR wascc_host::middleware::prometheus] Error registering gauge 'wascc:http_server_default_average_inv_time': Error: descriptor Desc { fq_name: "wascc:http_server_default_average_inv_time", help: "Average time to invoke operation \'BindActor\' on capability \'wascc:http_server\' with binding \'default\'", const_label_pairs: [], variable_labels: [], id: 2460217315795522843, dim_hash: 4187395188162649320 } already exists with the same fully-qualified name and const label values
```

**Comment by autodidaddict** [comment]
When I attempt to run the docker-compose command from the `prometheus.rs` module, I get the following output:

```
 docker-compose -f "docker-compose.yml" up -d --build
Creating network "promtest_default" with the default driver
Pulling prometheus (prom/prometheus:)...
Traceback (most recent call last):
  File "/usr/bin/docker-compose", line 11, in <module>
    load_entry_point('docker-compose==1.25.0', 'console_scripts', 'docker-compose')()
  File "/usr/lib/python3/dist-packages/compose/cli/main.py", line 72, in main
    command()
  File "/usr/lib/python3/dist-packages/compose/cli/main.py", line 128, in perform_command
    handler(command, command_options)
  File "/usr/lib/python3/dist-packages/compose/cli/main.py", line 1107, in up
    to_attach = up(False)
  File "/usr/lib/python3/dist-packages/compose/cli/main.py", line 1088, in up
    return self.project.up(
  File "/usr/lib/python3/dist-packages/compose/project.py", line 542, in up
    svc.ensure_image_exists(do_build=do_build, silent=silent, cli=cli)
  File "/usr/lib/python3/dist-packages/compose/service.py", line 360, in ensure_image_exists
    self.pull(silent=silent)
  File "/usr/lib/python3/dist-packages/compose/service.py", line 1248, in pull
    return progress_stream.get_digest_from_pull(event_stream)
  File "/usr/lib/python3/dist-packages/compose/progress_stream.py", line 102, in get_digest_from_pull
    for event in events:
  File "/usr/lib/python3/dist-packages/compose/service.py", line 1213, in _do_pull
    output = self.client.pull(repo, **pull_kwargs)
  File "/usr/lib/python3/dist-packages/docker/api/image.py", line 396, in pull
    header = auth.get_config_header(self, registry)
  File "/usr/lib/python3/dist-packages/docker/auth.py", line 47, in get_config_header
    authcfg = resolve_authconfig(
  File "/usr/lib/python3/dist-packages/docker/auth.py", line 324, in resolve_authconfig
    return authconfig.resolve_authconfig(registry)
  File "/usr/lib/python3/dist-packages/docker/auth.py", line 235, in resolve_authconfig
    cfg = self._resolve_authconfig_credstore(registry, store_name)
  File "/usr/lib/python3/dist-packages/docker/auth.py", line 262, in _resolve_authconfig_credstore
    store = self._get_store_instance(credstore_name)
  File "/usr/lib/python3/dist-packages/docker/auth.py", line 286, in _get_store_instance
    self._stores[name] = credentials.Store(
  File "/usr/lib/python3/dist-packages/docker/credentials/store.py", line 23, in __init__
    raise errors.InitializationError(
docker.credentials.errors.InitializationError: docker-credential-secretservice not installed or not available in PATH

```

**Comment by autodidaddict** [comment]
I tried opening both the prometheus and the grafana dashboards and neither of them registered any of the stats from waSCC. Also, is there a naming convention that can be used so that the wascc stats all appear as a category e.g. `go` or `scrape` etc?

I might just be running the example wrong (I ran the docker compose in one terminal and `cargo run --example prometheus_middleware --features "prometheus_middleware"` in the other)

grafana

**Comment by andrisak** [comment]
I've pushed a fix that should solve the error displayed when running the example. 

When you run the example, it should expose the metrics endpoint at `http://127.0.0.1:9898/metrics`. Can you browse or curl this url? On a get request it should display all the metrics in a simple text format.

Unfortunately, I still haven't solved the problem with running the examples on my Mac. Need to get it fixed so I can run everything properly.

**Comment by andrisak** [comment]
Good point about the grouping of values. I've prefixed all metrics with `wascc_`. This displays like this in Grafana:
Screenshot 2020-05-31 at 16 25 32
It only shows once the values are in Prometheus and Grafana has a configured data source pointing to `http://prometheus:9090`(assuming the configuration in the example).

What status does the scraping targets have in Prometheus? In the Prometheus web console, it's under Status > Targets.

I got the example `prometheus_middleware.rs` running now. This is what I get from `http://127.0.0.1:9898/metrics` when running the example:
```prometheus
# HELP wascc_actor_total_average_inv_time Average invocation time across all actors
# TYPE wascc_actor_total_average_inv_time gauge
wascc_actor_total_average_inv_time 0
# HELP wascc_actor_total_inv_count Total number of actor invocations
# TYPE wascc_actor_total_inv_count counter
wascc_actor_total_inv_count 0
# HELP wascc_cap_total_average_inv_time Average invocation time across all capabilities
# TYPE wascc_cap_total_average_inv_time gauge
wascc_cap_total_average_inv_time 0
# HELP wascc_cap_total_inv_count Total number of capability invocations
# TYPE wascc_cap_total_inv_count counter
wascc_cap_total_inv_count 2
# HELP wascc_wascc:http_server_default_BindActor_average_inv_time Average time to invoke operation 'BindActor' on capability 'wascc:http_server' with binding 'default'
# TYPE wascc_wascc:http_server_default_BindActor_average_inv_time gauge
wascc_wascc:http_server_default_BindActor_average_inv_time 0
# HELP wascc_wascc:http_server_default_BindActor_inv_count Number of invocations of operation 'wascc:http_server' on capability 'default' with binding 'BindActor'
# TYPE wascc_wascc:http_server_default_BindActor_inv_count counter
wascc_wascc:http_server_default_BindActor_inv_count 2
# HELP wascc_wascc:http_server_default_average_inv_time Average time to invoke capability 'wascc:http_server' with binding 'default'
# TYPE wascc_wascc:http_server_default_average_inv_time gauge
wascc_wascc:http_server_default_average_inv_time 0
# HELP wascc_wascc:http_server_default_inv_count Number of invocations of capability 'wascc:http_server' with binding 'default'
# TYPE wascc_wascc:http_server_default_inv_count counter
wascc_wascc:http_server_default_inv_count 2
```

**Comment by autodidaddict** [comment]
Turns out the reason why I couldn't get the Grafana and Prometheus apps to work is that `host.docker.internal` doesn't work on Linux. Do you think you could modify the testing instructions so that the docker-compose file works on both linux and mac, or at least contains instructions so that Mac and Linux users can both test it

**Comment by autodidaddict** [comment]
Confirmed that I can actually hit the metrics endpoint, so so the issue really is just in modifying the README so that linux folks can test.

@andrisak when the average invocation time shows `0.5` is that 0.5 milliseconds or 0.5 seconds?
```
❯ curl localhost:9898/metrics
# HELP wascc_MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2_HandleRequest_average_inv_time Average time to invoke operation 'HandleRequest' on actor 'MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2'
# TYPE wascc_MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2_HandleRequest_average_inv_time gauge
wascc_MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2_HandleRequest_average_inv_time 0.5
# HELP wascc_MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2_HandleRequest_inv_count Number of invocations of operation 'HandleRequest' on actor 'MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2'
# TYPE wascc_MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2_HandleRequest_inv_count counter
wascc_MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2_HandleRequest_inv_count 6
# HELP wascc_MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2_average_inv_time Average time to invoke actor 'MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2'
# TYPE wascc_MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2_average_inv_time gauge
wascc_MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2_average_inv_time 0.5
# HELP wascc_MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2_inv_count Number of invocations of actor 'MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2'
# TYPE wascc_MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2_inv_count counter
wascc_MB4OLDIC3TCZ4Q4TGGOVAZC43VXFE2JQVRAXQMQFXUCREOOFEKOKZTY2_inv_count 6
# HELP wascc_MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN_HandleRequest_average_inv_time Average time to invoke operation 'HandleRequest' on actor 'MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN'
# TYPE wascc_MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN_HandleRequest_average_inv_time gauge
wascc_MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN_HandleRequest_average_inv_time 0.2
# HELP wascc_MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN_HandleRequest_inv_count Number of invocations of operation 'HandleRequest' on actor 'MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN'
# TYPE wascc_MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN_HandleRequest_inv_count counter
wascc_MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN_HandleRequest_inv_count 5
# HELP wascc_MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN_average_inv_time Average time to invoke actor 'MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN'
# TYPE wascc_MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN_average_inv_time gauge
wascc_MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN_average_inv_time 0.2
# HELP wascc_MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN_inv_count Number of invocations of actor 'MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN'
# TYPE wascc_MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN_inv_count counter
wascc_MDFD7XZ5KBOPLPHQKHJEMPR54XIW6RAG5D7NNKN22NP7NSEWNTJZP7JN_inv_count 5
# HELP wascc_actor_total_average_inv_time Average invocation time across all actors
# TYPE wascc_actor_total_average_inv_time gauge
wascc_actor_total_average_inv_time 0.36363636363636365
# HELP wascc_actor_total_inv_count Total number of actor invocations
# TYPE wascc_actor_total_inv_count counter
wascc_actor_total_inv_count 11
# HELP wascc_cap_total_average_inv_time Average invocation time across all capabilities
# TYPE wascc_cap_total_average_inv_time gauge
wascc_cap_total_average_inv_time 0
# HELP wascc_cap_total_inv_count Total number of capability invocations
# TYPE wascc_cap_total_inv_count counter
wascc_cap_total_inv_count 2
# HELP wascc_wascc:http_server_default_BindActor_average_inv_time Average time to invoke operation 'BindActor' on capability 'wascc:http_server' with binding 'default'
# TYPE wascc_wascc:http_server_default_BindActor_average_inv_time gauge
wascc_wascc:http_server_default_BindActor_average_inv_time 0
# HELP wascc_wascc:http_server_default_BindActor_inv_count Number of invocations of operation 'wascc:http_server' on capability 'default' with binding 'BindActor'
# TYPE wascc_wascc:http_server_default_BindActor_inv_count counter
wascc_wascc:http_server_default_BindActor_inv_count 2
# HELP wascc_wascc:http_server_default_average_inv_time Average time to invoke capability 'wascc:http_server' with binding 'default'
# TYPE wascc_wascc:http_server_default_average_inv_time gauge
wascc_wascc:http_server_default_average_inv_time 0
# HELP wascc_wascc:http_server_default_inv_count Number of invocations of capability 'wascc:http_server' with binding 'default'
# TYPE wascc_wascc:http_server_default_inv_count counter
wascc_wascc:http_server_default_inv_count 2

```

**Comment by andrisak** [comment]
@autodidaddict That explains why no metrics ended up in Prometheus. It seems like there is no 1-1 equivalent to `host.docker.internal` on Linux. Here are some resources I found:

- [What is linux equivalent of “host.docker.internal”](https://stackoverflow.com/questions/48546124/what-is-linux-equivalent-of-host-docker-internal)
- [Support host.docker.internal DNS name to host](https://github.com/docker/for-linux/issues/264)

Does it work if you use `172.17.0.1` instead of `host.docker.internal`? If it does, that might be good enough for this type of documentation?

**Comment by andrisak** [comment]
@autodidaddict All times are in ms. I just pushed a commit where the help texts contain the unit of measure.

**Comment by autodidaddict** [comment]
Looks like 127.0.0.1 doesn't work for my on Ubuntu either. However, I can reliably get those stats from the metrics endpoint on the sample, so I'm okay with it being testable to this point, as the issue is just a matter of fussing with ports and networks, and not the actual prometheus middleware code.

**Review by autodidaddict — APPROVED** [review]

**Comment by amar123deep** [comment]
I can get the metrics from `http://127.0.0.1:9898/metrics`, but I couldn't get the metrics in grafana. I tried putting `172.17.0.1` in place of `host.docker.internal`, but that also didn't help. I was wondering if someone else has faced and solved the issue.

**Pull request merged**